File: extensions/h5/properties/FCPL.cc
// -*- C++ -*-
// -*- coding: utf-8 -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2026 all rights reserved


// externals
#include "external.h"
// namespace setup
#include "forward.h"


// file creation property lists
void
pyre::h5::py::properties::fcpl(py::module & m)
{
    // add bindings for hdf5 file creation property lists
    auto cls = py::class_<FCPL, PropList>(
        // in scope
        m,
        // class name
        "fcpl",
        // docstring
        "a file creation property list");

    // static properties
    cls.def_property_readonly_static(
        // the name
        "default",
        // the implementation
        [](const py::object &) -> const FCPL & {
            // easy enough
            return FCPL::theDefault();
        },
        // we hand back a reference to a shared, library-owned object
        py::return_value_policy::reference,
        // docstring
        "the default file creation property list");

    // constructor
    cls.def(
        // the implementation
        py::init(),
        // the docstring
        "build a file creation property list");

    // interface
    // the file space page size
    cls.def_property(
        // the name
        "pageSize",
        // the getter
        &FCPL::pageSize,
        // the setter
        &FCPL::setPageSize,
        // the docstring
        "the size of the pages my file space is carved into");

    // the file space strategy
    cls.def_property(
        // the name
        "filespaceStrategy",
        // the getter
        &FCPL::filespaceStrategy,
        // the setter, which unpacks the {(strategy, persist, threshold)} triplet
        [](FCPL & self,
           const std::tuple<H5F_fspace_strategy_t, bool, hsize_t> & strategy) -> void {
            // spread the triplet
            auto [approach, persist, threshold] = strategy;
            // and hand it to the property list
            self.setFilespaceStrategy(approach, persist, threshold);
            // all done
            return;
        },
        // the docstring
        "how free space is managed, as {(strategy, persist, threshold)}");

    // the user block
    cls.def_property(
        // the name
        "userblock",
        // the getter
        &FCPL::userblock,
        // the setter
        &FCPL::setUserblock,
        // the docstring
        "the size of the byte range at the front of my file that hdf5 leaves alone");

    // the widths used to record positions and lengths
    cls.def_property(
        // the name
        "sizes",
        // the getter
        &FCPL::sizes,
        // the setter, which unpacks the {(offset bytes, length bytes)} pair
        [](FCPL & self, const std::tuple<std::size_t, std::size_t> & sizes) -> void {
            // spread the pair
            auto [offsets, lengths] = sizes;
            // and hand it to the property list
            self.setSizes(offsets, lengths);
            // all done
            return;
        },
        // the docstring
        "the widths used to record positions and lengths, as {(offset, length)} bytes");

    // all done
    return;
}


// end of file

File: tests/h5.ext/fcpl.py
#!/usr/bin/env python3
# -*- Python -*-
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2026 all rights reserved


def test():
    """
    Exercise the file creation property list: the settings that are fixed when a file is
    made and cannot be revised afterwards
    """
    # get the bindings
    from pyre.extensions import libh5

    # the property namespace, which is where the records live
    p = libh5.properties

    # for a scratch path and its cleanup
    import os

    # a scratch data product
    uri = "h5_ext_fcpl.h5"
    # make sure a stale one is not lying around
    if os.path.exists(uri):
        os.remove(uri)

    # a fresh file creation property list
    fcpl = libh5.properties.fcpl()

    # the widths used to record positions and lengths; eight bytes each addresses more
    # than any product we will ever write
    assert (fcpl.sizes.offsets, fcpl.sizes.lengths) == (8, 8)

    # nothing is reserved at the front of the file, by default
    assert fcpl.userblock == 0
    # a file that another format has to carry reserves a block there; hdf5 requires it to
    # be a power of two, and at least 512 bytes
    fcpl.userblock = 1024
    assert fcpl.userblock == 1024

    # the free space strategy, as {(strategy, persist, threshold)}
    free = fcpl.filespaceStrategy
    # free space is not persisted across sessions, by default
    assert free.persist is False
    # ask for it to be, so a product that is revised repeatedly can reuse its own holes
    free.persist = True
    fcpl.filespaceStrategy = free
    assert fcpl.filespaceStrategy.persist is True

    # make a file with it
    f = libh5.File(uri=uri, mode="w", fcpl=fcpl)
    # and put something in it, so the file is worth reading back
    f.create(path="group")

    # all done
    return


# main
if __name__ == "__main__":
    # drive
    test()


# end of file
